### PR TITLE
Enable log record transformations

### PR DESCRIPTION
**Pull request by psarakis**
### Changes proposed:

- [x] Add log record transformations, including serialization and enrichment.
- [x] Add unit tests using Application Load Balancer log sample.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -44,18 +62,63 @@ All source code and dependencies can be found on the [le_lambda Github page](htt
...
+  * You can perform transformation operations for each log record, such as:
+    - Enrich the log record with the URL components or S3 path components (usually the `prefix`).
+    - Serialize to KVP format, instead of JSON which is the default.
+  * Transformations are applied sequentially and the entire pipeline can be defined via environment variable.
+
+
+  | Key | Value|
+  |------|------|
+  | `TRANSFORMER_CLASS_LIST` | Comma-separated list of transformer classes. |
+  | `TRANSFORMER_S3_KEY_FIELD_EXTRACTOR_MAPPING` | JSON-encoded field value mapping for the S3 path components enrichment: `[{"field": "my_field_1", "value": "key[0]"]`|
```
Comment (fardelas) [inline README.md]: I noticed a missing `}` at the end of the array

File: README.md [hunk 2]
```diff
@@ -44,18 +62,63 @@ All source code and dependencies can be found on the [le_lambda Github page](htt
...
+  The field-value mapping is specified using an array of objects:
+
+  ```
+  [{
+    "field": "<NAME_OF_ADDED_FIELD>",
+    "value": "key<SUBPATH_INDEX>"
+  }]
+  ```
+
+  For example, defining `{"field": "my_field", "value": "key[1]"'}` and the S3 key is: `/prefix1/prefix2/AWSLogs/`
```
Comment (fardelas) [inline README.md]: There is an extra `'` at the end of the dict

File: tests/test_le_lambda.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    def test_invocation(self, validate_uuid, create_socket, s3):
+        _ = validate_uuid
+        mock_socket = create_socket.return_value
+        mock_socket.sendall = mock.MagicMock()
+        s3_contents = StringIO()
+        with gzip.GzipFile(fileobj=s3_contents, mode="w") as f:
+            f.write(
+                'https 2017-09-12T18:49:38.910213Z app/'
+                '22ff55229966cc00aa9911eeaa226677/9999cceedd225588 '
+                '254.254.254.5:8750 10.0.3.100:32831 0.001 0.002 0.000 '
```
Comment (fardelas) [inline tests/test_le_lambda.py]: There are two corner cases here regarding the IP and port parsing:
a. the field that contains the target ip and port may contain a `-`, which will break the parsing as it is implemented in le_lambda.py right now
b. IPv6 addresses both for clients and targets cannot get parsed correctly based on the current parsing implementation

Comment (psarakis) [reply tests/test_le_lambda.py]: `-` is handled now for ALB logs and test case has been added.

**Review by fardelas — APPROVED** [review]
I like!
